Barillon_Terry.cpp

```diff
@@ -0,0 +1,176 @@
+#include <iostream>
+#include <vector>
+#include <cstdlib>
+#include <ctime>
+using namespace std;
+
+// écrire une fonction permettant de calculer le nom d'un Zombie
+// prend en paramètres un string, retourne un string
+
+string calculer_nom_zombie(string nom_humain) {
+    int i;
+    string nom_zombie = "ntr";
+    for (i=0; i!=3; i=i+1)  {
+        nom_zombie[i] = nom_humain[0];
+    }
+    return string(nom_zombie);
+}
+
+// écrire une classe Humain avec un attribut nom, et un constructeur prenant un paramètre pour le nom.
+// attribut privé nom-string / accesseur/getter
+// méthode afficher()
+
+class Humain {
+private:
+    string nom;
+
+public:
+    Humain(string nom) {
+        this->nom = nom;
+    }
+
+    string getNom() {
+        return nom;
+    }
+
+    void afficher() {
+        cout << "Je suis un Humain, mon nom est " << nom << "." << endl;
+    }
+};
+
+//
+// écrire une classe zombie avec un attribut nom, et un constructeur prenant un Humain en paramètre
+//
+
+class Zombie {
+private:
+    string nom;
+
+public:
+    Zombie(Humain* humain) {
+        nom = "Zombie " + calculer_nom_zombie(humain->getNom());
+    }
+
+    Zombie* manger(Humain* humain) {
+        cout << nom << " mange " << humain->getNom() << endl;
+        return new Zombie(humain);
+    }
+
+    void afficher() {
+        cout << nom << endl;
+    }
+
+    string getNom() {
+        return nom;
+    }
+};
+
+//
+//exercice 4
+//
+
+class FinDuMonde {
+private:
+    vector<Humain*> humains;
+    vector<Zombie*> zombies;
+public:
+    void ajouterHumain(Humain* humain) {
+        // ajouter un humain
+        humains.push_back(humain);
+    }
+
+    void debutDeLaFin(Zombie, Humain) {
+        int index = rand() % humains.size();
+        Humain* humain = humains[index];
+
+        Zombie* zombie = new Zombie(humain);
+        zombies.push_back(zombie);
+
+        humains.erase(humains.begin() + index);
+
+        cout << "Le premier Zombie est " << zombie->getNom() << "." << endl;
+    }
+
+    void faireUnPas() {
+        for (int i = 0; i < zombies.size(); i++) {
+            // Pour autant de zombies présents
+            Zombie* zombie = zombies[i];
+
+            if (humains.empty()) {
+                // permet de savoir si il reste des humains ou non.
+                cout << "Il n'y a plus d'humains sur Terre, la fin du monde est arrivée" << endl;
+            }
+
+            else {
+                // Nous choisissons un humain au hasard dans la liste car pas de jaloux
+                int index = rand() % humains.size();
+                Humain* humain = humains[index];
+
+                // Il se fait mordre et devient donc un zombie, lui permettant de rentrer dans la liste
+                zombies.push_back(zombie->manger(humain));
```

Comment:
Attention : on ajoute des éléments dans le tableau d'itération...
Idéalement pour limiter les effets étonnants, on utilisera un deuxième tableau (un clone) pour itérer.